### DISCUSSION TITLE
Add translation with A.I service ChatGPT

### DISCUSSION THREAD

**Discussion by McFL7** [dimkroon/translate_subs#9] [Ideas]
Hi,

Great idea for subtitle translation. I was looking for an addon like yours. However it will be good to have your addon being able to connect to a paid A.I service API like ChatGPT, Gemini, Bing, Claude to translate subs to any language, as Google Translate is very bad on that matter. 

I am seeing big opportunity of improvement if your addon can send local or downloaded subtitles by another addon like A4ksubtitles or Subscene, to these A.I platforms and beijg automatically translated. as most of the time i am not finding french subtitles for my tv shows or movies. only english subs are available.

Imo it's a good feature to add to your addon.

**Comment by dimkroon** [comment]
Hi McFl7,
Many thanks for your comment and congratulations with the very first post in the discussions section!

The whole access to online translators is done by translatepy. I just pack that project in an add-on for use in Kodi. Although I've no principle objections to writing  such modules myself, I think it's best to first direct requests for access to other translation service to them. However, in order to develop and maintain access to paid servers I must have paid accounts to these services. Unless people are going to sponsor me this not going happen, and I'm afraid that the devs at translatepy don't think much differently.

I've never used a paid translation service, but I always believed that the main difference with the free service  was the limit on the number of translated tokens, rather than the quality of the translation. FAIK most, if not all modern translation services are A.I. powered. Google Translate certainly is. Funny you mentioning Google Translate being very bad. In the previous version of the add-on I started to specifically select Google V2 because it produced the best result. Do you perhaps have an example of a subtitle translation done by a paid service v.s. one by Google Translate? I'd be very interested to see what the differences are.

TranslateSubs was initially developed with the intention to translate subtitles files provided by video add-ons, actually subtitles for the hearing impaired to be more specific. As such, video add-ons must specifically implement support for TranslateSubs. If the video add-on already provides subtitles it's usually very simple to do. If you use a video add-on to play these TV shows and films I thinks it's be best to try to modify this add-on to support TranslateSubs. Maybe I should write a small howto for people who want to modify an add-on, or add-on developers even, who want to implement translation support.

Translating subtitles from other sources is something I'll have to look into. TBH, I don't use subtitle add-ons myself and I must admit that I have no idea how users normally interact with such add-ons and how to fit in translation. All automatic will be ideal of course, but first I'll have to figure out how subtitle add-ons interact with Kodi and vice versa. 
As a start, I could add the possibility to browse the local file system for a subtitle file and do a manual translation. Requires quite a bit of user interaction and lacks any kind of elegance, but is relatively easy to implement. Would that be a workable solution do you think?

**Reply by McFL7** [reply]
Hi,

I tried a translation with my paid ChatGPT account. it don't seem to want to translate the whole SRT file. Maybe a dedicated A.I service for translation will work. Yeah, Being able to load a srt file from a local folder or from a subtitle addon and save a translated version of the file in the same folder could be a interesting addition.

Could you implement it ? I will be happy to test it.

Great work.

**Reply by dimkroon** [reply]
Translating just a plain SRT file usually has very poor results. My work in the add-on is primarily focused on finding the best way to offer subtitles text to the translation service. 
Full sentences translate much better than the short phrases found in SRT's. I was kinda hoping that translation services translate lines in the context of the whole text, but I haven't seen any evidence of that so far. Maybe it's because of the way I offer the text, but perhaps a paid A.I. service does a better job in that respect.

Great that you are willing to test; I could definitely use some feedback. I'll try to add the possibility to translate local files. I just have quite a lot on my hands lately, so it may take a while. I'll drop a note here when I'm ready. 
Just let me know if you use video add-ons that produce subtitles you want to translate. It's usually a very simple modification to support automatic translation.

**Reply by McFL7** [reply]
I use mostly those addons : A4Ksubtitles, Opensubtitles.com, Opensubtitles.org, Podnapsi, Subscene. They don't always find French subs so want to have a way to choose the English subs they find and automaticaly translate them in French (or any other language for a more general usage).

**Comment by dimkroon** [comment]
@McFL7 
So, I finally got a first version of manual translation working. You can download the pre-release from https://dimkroon.net/kodi-addons/service.subtitles.translate_0.4.0-nightly-1.zip.

TranslateSubs registers itself now also as a subtitle add-on, so you can find it in the regular subtitle download dialog. Since this add-on does not search for downloadable subtitles on a web service it works slightly different to normal subtitle add-ons.

If you select TranslateSub in the subtitle download dialog it display's the subtitles found on your system that matches the playing video. The flag on each item represents the language of the subtitle file. The language text below the flag is language the subtitle will be translated to, which should be your preferred subtitle language. 
If you do not have a subtitle file, try to use a normal subtitle add-on to downlaod one and then try TranslateSubs again. 
If you are confident that you have a subtitle file, but the file is not listed, you can select the option 'Manually search for ...', which will open a file browser dialog, enabling you to manually select the file.

TranslateSubs works internally on srt files. It has a pretty decent converter for webvtt files and a basic converter for files in ttml format. Other subtitle formats are currently not supported.

Please give it a go, I'd very interested to hear if and how this works for you.

**Reply by McFL7** [reply]
Ohh great work and achievement 👍👏👏

I can't wait to test it.

Will get back to you with my test results.

**Comment by dimkroon** [comment]
Any news @McFL7?

**Reply by McFL7** [reply]
Hi,

It seems there is some issue. In the addon the target language is defined by the prefered ones in the player>subtitles settings. However it should not, as i have two prefered languages French and Englishs subs. And the addon asks me when i want to translate and English sub that its is already in the prefered languages, however its not the only prefered, how to force it select the other prefered language French. And when i change player settings to French, other subs addons don't list english one and i can't download them anymore to translate them with your addon. It seems the snake bits its tail. The target language should be set in the addons owns settings and not grabbed from player settings

A part from that when i selected an English sub file to translate, the addon showed me an error notification to download the subtitle file (while it's local, strange!!). see joined captures

![IMG_20240629_011741.jpg](https://github.com/dimkroon/translate_subs/assets/98433855/fcaba79f-59e8-4c5c-9171-636cc2677a7c)

![IMG_20240629_011751.jpg](https://github.com/dimkroon/translate_subs/assets/98433855/ed49417f-9d35-40f6-9720-78058476f4f7)

Regards

**Comment by dimkroon** [comment]
Many thanks for your feedback @McFL7!

I don't quite understand how you can have 2 preferred subtitle languages. In Settings -> Player -> Language - >Subtitle you can only select one language. And that's logical. Except for the bilingual person who really doesn't care at all, there will always one language that you'd rather have than any another. 

If I understand correctly, what you actually mean is that despite your preference for French you'd like to be able to download both French and English subtitles. But that's a completely different settings in Settings -> Player -> Subtitles, section 'Download Services'. The setting 'Languages to download subtitles for' allows you to select multiple languages. Subtitle add-ons should respect this setting and offer subtitles in all languages selected there. 

I strongly believe that user should be able to enter their preferences only in one place.  I very much like to avoid another preferred language setting in the add-on.
Can you check your the settings, possibly try various subtitle add-on and let me know if that works? I used opensubtitles.com to test and that worked fine.

As to the error notification:
I don't speak French, so the screenshot doesn't tell me much, but the notification alone would probably not provide sufficient info anyway. Could you replicate the error and send us a debug log? (Check https://kodi.wiki/view/Log_file on how to create and upload a log file)

**Reply by McFL7** [reply]
Hi,

Quick feedback

You are right there is only French as preferred language. I was confusing with the setting which allows to select multiple languages for subtitles search.


However, Still getting the same error, "unable to download subtitle" after selecting a sub file that was downloaded by another addon (a4ksubtitles or opensubtitles). 

I select the sub file in the subs folder with another addon, i select it then with Translatesubs and i get "error downloading subtitle".

After i select, It is also always telling me the subtitle is already in the translated language while it is not, it is always an English sub i want to translate to French.
